### PR TITLE
Update Speech-to-Text

### PR DESCRIPTION
**Pull request by da03**
Add sharded data loader for processing large speech2text corpus. Model changed to Baidu DeepSpeech 3 seq2seq-attn (https://arxiv.org/abs/1707.07413).

### CODE REVIEW

File: onmt/io/AudioDataset.py [hunk 1]
```diff
@@ -2,10 +2,12 @@
...
 import codecs
 import os
+import io
+from tqdm import tqdm
 
 import torch
 import torchtext
-
+import sys
 from onmt.io.DatasetBase import ONMTDatasetBase, PAD_WORD, BOS_WORD, EOS_WORD
```
Comment (JianyuZhan) [inline onmt/io/AudioDataset.py]: I would follow the practice of import order:  system, then third-party, and finally inner-project.

File: onmt/io/AudioDataset.py [hunk 2]
```diff
@@ -35,20 +37,10 @@ class AudioDataset(ONMTDatasetBase):
     """
```
Comment (JianyuZhan) [inline onmt/io/AudioDataset.py]: Stale comments to update?

File: preprocess.py [hunk 3]
```diff
@@ -117,6 +117,74 @@ def build_save_text_dataset_in_shards(src_corpus, tgt_corpus, fields,
...
+    easily reclaim these memory.
+
+    If `max_shard_size` is 0 or is larger than the corpus size, it is
+    effectively preprocessed into one dataset, i.e. no sharding.
+
+    NOTE! `max_shard_size` is measuring the input corpus size, not the
+    output pt file size. So a shard pt file consists of examples of size
+    2 * `max_shard_size`(source + target).
+    '''
+
```
Comment (JianyuZhan) [inline preprocess.py]: Stale comment to update

File: onmt/io/AudioDataset.py [hunk 4]
```diff
@@ -283,3 +290,122 @@ def get_num_features(corpus_file, side):
...
+                        "Two corpuses must have same number of lines!")
+
+                self.line_index += 1
+                iteration_index += 1
+                yield self._example_dict_iter(line, iteration_index)
+
+            if self.assoc_iter.eof:
+                self.eof = True
+                self.corpus.close()
+        else:
```
Comment (JianyuZhan) [inline onmt/io/AudioDataset.py]: The original associate iterator is to synchronize the shards between source text file and target  text file. Concretely, we use `max_shard_size` to determine how many lines a shard of source text file contains, then make the target side synchronize with it, using the `assoc_iter`.

IIUC,  for Audio corpus, the source side is a specific audio file, thus the target side shall be the corresponding target text lines, so I think the `assoc_iter` logic should change to reflect this fact?

Also stale comment to update.

**Comment by da03** [comment]
Okay ready to be reviewed.

**Comment by vince62s** [comment]
@da03 is there any benchmark on popular datasets for this speech-2-text thing ? (WSJ or tedlium)

**Comment by da03** [comment]
Performance-wise, this achieves WER of 21.6% on Librispeech full (clean+other) test set (Experiments by Thierry Tambe @ttambe) with batch size 8, rnn size 512, 4 encoder layers and 1 decoder layer, adam optimizer and character level outputs. In comparison, a comparable model (Zeyer et al, "Improved training of end-to-end attention models for speech recognition", no outside LM) gets 4.87% | 15.89% on clean | other with pretraining, BPE'ed outputs and 6 layer encoder.

**Comment by bpopeters** [comment]
Yep, that's where the conflict comes from, but the fix is just common sense.

**Comment by bpopeters** [comment]
First, I think this pull request definitely improves the codebase and adds necessary features. I haven't looked at it in too much depth besides the sharded iterator and the place where it conflicted with the sparsemax loss stuff, but it should probably be merged.

However, I do have a question and some more general comments about how we approach data processing and different data types.

Is there anything about the `ShardedAudioCorpusIterator` that is genuinely audio-specific? It looks like the implementation of `_example_dict_iter` is different, but this could be done in a way that doesn't require so much reduplication of code.

In general, I think most of the data processing pipeline does not need to be modality-specific. The framework that pytorch provides us has three basic parts:

1.  The fields describe how raw data is preprocessed, including tokenization and numericalization and other transformations you might want to apply.
2. A dataset applies the first part of the fields' operations to the data to create examples (so the data is cleaned but not necessarily numericalized yet).
3. An iterator takes the examples stored in the dataset and creates tensors, sorting and batching as necessary.

Very little of this process is modality-specific: different data needs to be tokenized in different ways, but the rest is general. Writing modality-specific functions makes the codebase unfriendly to people trying out new forms of input because of all the code that needs to be rewritten (often in several places). And it gets worse if you relax the constraint that the data on the target side is always text. This is something our data processing assumes: can you imagine writing separate dataset and iterator classes for each _pair_ of (src, tgt) datatypes? What about if you have a multimodal dataset? Clearly these are places where it helps to be more general.

File: onmt/encoders/image_encoder.py [hunk 5]
```diff
@@ -16,11 +16,12 @@ class ImageEncoder(nn.Module):
         dropout (float): dropout probablity.
     """
 
-    def __init__(self, num_layers, bidirectional, rnn_size, dropout):
+    def __init__(self, num_layers, bidirectional, hidden_size, dropout):
```
Comment (vince62s) [inline onmt/encoders/image_encoder.py]: I think you need to remove all the "image" related changes since I merged the PR fixing all of this + sharding based on number of examples.
Your PR shoudl handle only Speec-2-Text IMO for now.

File: onmt/inputters/image_dataset.py [hunk 6]
```diff
@@ -173,7 +173,7 @@ def get_fields(n_src_features, n_tgt_features):
         fields = {}
 
         def make_img(data, vocab):
-            """ ? """
+            """ batch img data """
```
Comment (vince62s) [inline onmt/inputters/image_dataset.py]: same comment, let's remove the image changes from this PR

**Comment by da03** [comment]
Okay thanks for the feedback! Working on it.

**Comment by da03** [comment]
Okay ready to be reviewed. I think Ben is making a good point about reusing data loaders, but this PR mainly intends to fix the speech model, and we can leave code restructuring to another PR.

File: onmt/models/model.py [hunk 7]
```diff
@@ -42,7 +43,10 @@ def forward(self, src, tgt, lengths, dec_state=None):
         """
         tgt = tgt[:-1]  # exclude last target from inputs
 
-        enc_final, memory_bank = self.encoder(src, lengths)
+        if self.model_type == 'audio':
+            enc_final, memory_bank, lengths = self.encoder(src, lengths)
```
Comment (vince62s) [inline onmt/models/model.py]: instead of making a specific case, why don't make all encoder interfaces the same?

**Comment by da03** [comment]
Okay, ready to be reviewed. Device argument has been deleted, and all models use the same interface to get rid of model_type in model class.

**Comment by vince62s** [comment]
looks good to me.
Just one thing: do you think the new shard_size based on examples would work for audio too ? (what Viet did for the image stuff)
if so, I think we could just keep this way, what do you think ?

**Comment by da03** [comment]
Ok fixed, now audio sharding uses shard_size as in image case.

**Comment by vince62s** [comment]
okay, great, merging. I think we will get rid of the max_shard_size for text too. ok ?

**Comment by da03** [comment]
Yeah sounds good.

**Comment by Zer0-dev115** [comment]
How Can I use CTC with RNN in OpenNMT for Speech to text?

**Comment by Zer0-dev115** [comment]
> How Can I use CTC with RNN in OpenNMT for Speech to text?

Currently I am using BRNN for libri-speech data.
